README.md
# 企业微信机器人接入

|                           | [Linux+OSX][linux-link] | [Windows MSVC+GNU][windows-link] |
|:-------------------------:|:-----------------------:|:--------------------------------:|
| Build & Publish           | ![linux-badge]          | ![windows-badge]                 |

[linux-badge]: https://travis-ci.org/owt5008137/wxwork_robotd.svg?branch=master "Travis build status"
[linux-link]:  https://travis-ci.org/owt5008137/wxwork_robotd "Travis build status"
[windows-badge]: https://ci.appveyor.com/api/projects/status/ht5pks682ehe2vkt?svg=true "AppVeyor build status"
[windows-link]:  https://ci.appveyor.com/project/owt5008137/wxwork-robotd "AppVeyor build status"

来自企业微信机器人的请求会先按填入的URL的最后一节匹配到配置中 ```projects``` 里 ```name``` 对应的项目中，然后优先从 ```projects``` 内尝试匹配命令，再从全局域匹配命令。 命令的类型( ```type``` )支持 **echo （输出消息）** 、 **http （HTTP请求）** 、 **help （帮助信息）** 和 **spawn （子进程执行命令）** 。 配置项中的 *echo* 、 *exec* 、 *args* 、 *url* 、 *post* 、 *prefix* 、 *suffix* 会使用模板引擎尝试对内容进行替换（传参），传入的参数可以是匹配命令里的匹配结果，也可以是预先配置好的环境变量。

如果是 **spawn （子进程执行命令）** 类型的请求，配置里的环境变量在子进程中也可用。

可以通过 ```./wxwork_robotd -h``` 查看可用选项。[etc](etc) 目录里有各种系统的启动脚本模板。


## 下载和安装

您可以在 https://github.com/owt5008137/wxwork_robotd/releases 下载预发布包，解压即可。

> 或者使用 rust 的cargo 命令```cargo install wxwork_robotd``` 来直接安装到 ```$HOME/.cargo/bin``` 。
> 这种方式只会安装可执行程序，其他的示例文件和工具脚本可以在 ```$HOME/.cargo/registry/src/github.com-*/wxwork_robotd-*``` 里找到。

发布包文件说明：

1. [etc/conf.json](etc/conf.json) ：示例的配置文件。
2. [etc/firewalld/wxwork_robotd.xml](etc/firewalld/wxwork_robotd.xml) ： 示例的firewalld配置文件。
  > 请先修改端口号为和配置文件保持一致。
  > 然后复制到 ```/etc/firewalld/services/wxwork_robotd.xml``` 后执行 ```sudo firewall-cmd --permanent --add-service=wxwork_robotd``` 即可。

3. [etc/systemd/wxwork_robotd.service](etc/systemd/wxwork_robotd.service) ： 示例的systemd服务配置文件
  > 请先修改部署目录和实际使用的路径保持一致
  > 然后复制到 ```/usr/lib/systemd/system/wxwork_robotd.service``` 后执行 ```sudo systemctl enable wxwork_robotd && sudo systemctl start wxwork_robotd``` 即可。

4. [etc/systemv/wxwork_robotd](etc/systemv/wxwork_robotd) ： 示例的用于 systemv 的服务配置文件
5. [etc/init.d/wxwork_robotd](etc/init.d/wxwork_robotd) ： 示例的用于 init.d 的服务配置文件
6. [tools](tools) ： 用于主动发机器人消息的工具脚本


## 主动发消息接口

[tools](tools) 目录包含用于主动发消息的**企业微信机器人脚本**，兼容 python 2.7-3.X 。 

> python 2.7 依赖 [requests](https://pypi.org/project/requests/)库。 可通过 ```pip install requests``` 来安装。

## 环境变量命名

+ CMD和匹配的内容可以通过 ```{{WXWORK_ROBOT_CMD}}``` 和 ```{{WXWORK_ROBOT_CMD_<变量名或匹配名>}}``` 来获取。
+ ```projects``` 中的内容可以通过 ```{{WXWORK_ROBOT_PROJECT_<变量名或匹配名>}}``` 来获取。
+ 环境变量只会导出类型为字符串、数字或者布尔值的内容，不支持嵌套内容
+ 可用的环境变量
  * WXWORK_ROBOT_VERSION                    : 当前机器人版本号(版本>=0.10.2)
  * WXWORK_ROBOT_WEBHOOK_KEY                : 当前消息对应机器人的WebhookURL里的key字段(可用来回发消息,版本>=0.3.3)
  * WXWORK_ROBOT_WEBHOOK_URL                : 当前消息对应机器人的WebhookURL(可用来回发消息,版本>=0.3.3)
  * WXWORK_ROBOT_CMD                        : 当前执行命令的完整匹配消息
  * WXWORK_ROBOT_CMD_{VARNAME}              : 当前执行命令的匹配参数（必须是命名匹配式）或配置的环境变量
  * WXWORK_ROBOT_PROJECT_NAME               : 配置的项目名
  * WXWORK_ROBOT_PROJECT_TOKEN              : 配置的项目验证token
  * WXWORK_ROBOT_PROJECT_ENCODING_AES_KEY   : 配置的项目base64的aes key
  * WXWORK_ROBOT_PROJECT_{VARNAME}          : 配置的项目中的环境变量
  * WXWORK_ROBOT_MSG_FROM_USER_ID           : 发消息者的用户id(版本>=0.3.6)
  * WXWORK_ROBOT_MSG_FROM_NAME              : 发消息者的用户名称(版本>=0.3.6)
  * WXWORK_ROBOT_MSG_FROM_ALIAS             : 发消息者的用户别名(版本>=0.3.6)
  * WXWORK_ROBOT_MSG_ID                     : 消息ID(版本>=0.3.6)
  * WXWORK_ROBOT_IMAGE_URL                  : 如果是图文混排和图片消息，这个指向消息内的图片地址(版本>=0.8.0)
  * WXWORK_ROBOT_GET_CHAT_INFO_URL          : 可以用于获取消息信息的URL(版本>=0.3.9)，有效期为5分钟，调用一次后失效
  * WXWORK_ROBOT_POST_ID                    : post id(版本>=0.9.0)
  * WXWORK_ROBOT_CHAT_ID                    : chat id(版本>=0.3.9)，用于区分聊天群，如果机器人被添加到多个群，可以用这个指定主动发消息到哪个群
  * WXWORK_ROBOT_CHAT_TYPE                  : chat type(版本>=0.6.1)，对应企业微信机器人消息的ChatType字段（会话类型，single/group，分别表示：单聊\群聊话）
  * WXWORK_ROBOT_HTTP_RESPONSE              : HTTP回包(仅 ```type``` 为 http 时的echo字段可用)
  * WXWORK_ROBOT_MSG_TYPE                   : msg type(版本>=0.7.0)，对应企业微信机器人消息的MsgType字段（text/event/attachment）
  * WXWORK_ROBOT_APP_VERSION                : msg type(版本>=0.7.0)，对应企业微信机器人消息的AppVersion字段
  * WXWORK_ROBOT_EVENT_TYPE                 : msg type(版本>=0.7.0)，对应企业微信机器人消息的EventType字段（目前可能是add_to_chat表示被添加进群，或者delete_from_chat表示被移出群, enter_chat 表示用户进入机器人单聊）
  * WXWORK_ROBOT_ACTION_NAME                : msg type(版本>=0.7.0)，对应企业微信机器人消息的Actions.Name字段（用户点击按钮的名字）
  * WXWORK_ROBOT_ACTION_VALUE               : msg type(版本>=0.7.0)，对应企业微信机器人消息的Actions.Value字段（用户点击按钮的值）
  * WXWORK_ROBOT_ACTION_CALLBACKID          : msg type(版本>=0.7.0)，对应企业微信机器人消息的Attachment.CallbackId字段（attachment中设置的回调id）


## 配置说明

注意，下面只是配置示例，实际使用的配置必须是标准json，不支持注释

```javascript
{
    "listen": ["0.0.0.0:12019", ":::12019"], // 监听列表，这里配置了ipv4和ipv6地址
    "task_timeout": 4000,                    // 超时时间4000ms，企业微信要求在5秒内回应，这里容忍1秒钟的网络延迟
    "workers": 8,                            // 工作线程数
    "backlog": 256,                          // 建立连接的排队长度
    "keep_alive": 5,                         // tcp保持连接的心跳间隔（秒） (版本: >=0.6.0)
    "client_timeout": 5000,                  // 客户端第一个请求的超时时间（毫秒） (版本: >=0.6.0)
    "client_shutdown": 5000,                 // 客户端连接的超时时间（毫秒） (版本: >=0.6.0)
    "max_connection_per_worker": 20480,      // 每个worker的最大连接数，当连接数满之后不会再接受新连接 (版本: >=0.6.0)
    "max_concurrent_rate_per_worker": 256,   // 每个worker的最大握手连接数，当连接数满之后不会再接受新连接（一般用于控制SSL握手的开销） (版本: >=0.6.0)
    "payload_size_limit": 262144,            // 消息体最大长度，默认: 262144(256KB) (版本: >=0.6.0)
    "static_root": ".",                      // 静态文件服务根目录，默认不开启 (版本: >=0.10.1)
    "cmds": {                                // 这里所有的command所有的project共享
        "default": {                         // 如果找不到命令，会尝试找名称为default的命令执行，这时候
            "type": "echo",                  // 直接输出类型的命令
            "echo": "我还不认识这个指令呐!({{WXWORK_ROBOT_CMD}})", // 输出内容
            "order": 999,                    // 命令匹配优先级，越小则越优先匹配，默认为 0
            "hidden": true                   // 是否隐藏，所有的命令都有这个选项，用户help命令隐藏这条指令的帮助信息
        },
        "": {                               // 如果输入了空消息或者attachment消息，则会匹配这个命令而不是default,没有配置空命令则会直接忽略输入
            "type": "echo",
            "echo": "Hello, 本群会话ID: {{WXWORK_ROBOT_CHAT_ID}}",
            "order": 999,
            "hidden": true
        },
        "(help)|(帮助)|(指令列表)": {
            "type": "help",                    // 帮助类型的命令
            "description": "help|帮助|指令列表", // 描述，所有的命令都有这个选项，用于help类型命令的输出，如果没有这一项，则会直接输出命令的key（匹配式）
            "prefix": "### 可用指令列表\r\n"     // 帮助信息前缀
            "suffix": ""                       // 帮助信息后缀
            "case_insensitive": true,          // [所有命令] 是否忽略大小写（默认:true）
            "multi_line": true,                // [所有命令] 是否开启逐行匹配（默认:true，When enabled, ^ matches the beginning of lines and $ matches the end of lines.）
            "unicode": true,                   // [所有命令] 是否开启unicode支持（默认:true，When disabled, character classes such as \w only match ASCII word characters instead of all Unicode word characters）
            "octal": true,                     // [所有命令] 是否支持octal语法（默认:false）
            "dot_matches_new_line": false,     // [所有命令] .是否匹配换行符（默认:true）
            "order": 0                         // [所有命令] 命令匹配优先级，越小则越优先匹配(默认: 0)
        },
        "说\\s*(?P<MSG>[^\\r\\n]+)": {
            "type": "echo",
            "echo": "{{WXWORK_ROBOT_CMD_MSG}}", // 可以使用匹配式里的变量
            "description": "说**消息内容**",
            "order": 2
        },
        "执行命令\\s*(?P<EXEC>[^\\s]+)\\s*(?P<PARAM>[^\\s]*)": {
            "type": "spawn",                    // 启动子进程执行命令，注意，任务超时并不会被kill掉
            "exec": "{{WXWORK_ROBOT_CMD_EXEC}}",
            "args": ["{{WXWORK_ROBOT_CMD_PARAM}}"],
            "cwd": "",
            "env": {                            // 命令级环境变量，所有的命令都有这个选项，这些环境变量仅此命令有效
                "TEST_ENV": "all env key will be WXWORK_ROBOT_CMD_{NAME IN ENV} or WXWORK_ROBOT_PROJECT_{NAME}"
            },
            "description": "执行命令**可执行文件路径** ***参数***",
            "output_type": "输出类型",          // markdown/text
            "order": 2
        }
    },
    "events": {                                                             // 这里的事件所有project共享
        "add_to_chat": {                                                    // 加入群聊（内容和命令一样）
            "type": "echo",
            "echo": "Hi, 大家好"
        },
        "enter_chat": {                                                     // 加入单聊（内容和命令一样）
            "type": "echo",
            "echo": "Hi, {{WXWORK_ROBOT_MSG_FROM_NAME}}"
        }
    },
    "projects": [{                                                          // 项目列表，可以每个项目对应一个机器人，也可以多个机器人共享一个项目
        "name": "test_proj",                                                // 名称，影响机器人回调路径，比如说这里的配置就是: http://外网IP:/12019/test_proj/
        "token": "hJqcu3uJ9Tn2gXPmxx2w9kkCkCE2EPYo",                        // 对应机器人里配置的Token
        "encodingAESKey": "6qkdMrq68nTKduznJYO1A37W2oEgpkMUvkttRToqhUt",    // 对应机器人里配置的EncodingAESKey
        "env": {                                                            // 项目级环境变量，这些环境变量仅此项目有效
            "testURL": "robots.txt"
        },
        "cmds": {                                                           // 项目级命令，这些命令仅此项目有效
            "http请求": {
                "type": "http",                                             // http请求类命令
                "method": "get",                                            // http方法，可选值为 get/post/put/delete/head，如果不填则会自动从设置，如果post里有数据则会自动设为post，否则自动设为get
                "url": "https://owent.net/{{WXWORK_ROBOT_PROJECT_TEST_URL}}", // http请求地址
                "post": "",                                                   // body里的数据
                "content_type": "",                                           // content-type，可不填
                "headers": {                                                  // 请求的额外header
                    "X-TEST": "value"
                },
                "echo": "已发起HTTP请求，回包内容\r\n{{WXWORK_ROBOT_HTTP_RESPONSE}}" // 机器人回应内容
            },
            "访问\\s*(?P<URL>[^\\r\\n]+)": {
                "type": "http",
                "url": "{{WXWORK_ROBOT_CMD_URL}}",
                "post": "",
                "echo": "HTTP请求: {{WXWORK_ROBOT_CMD_URL}}\r\n{{WXWORK_ROBOT_HTTP_RESPONSE}}",
                "description": "访问**URL地址**"
            }
        },
        "events": {                                                           // 这里的事件仅当前project有效
            "delete_from_chat": {                                             // 离开群聊
                "type": "echo",
                "echo": "再见"
            }
        }
    }]
}
```

## 语法相关

+ 完整示例配置见 [etc/conf.json](etc/conf.json)
    > 请确保 taskTimeout 字段低于5000毫秒，因为企业微信的超时是5秒，如果加上网络延迟之后机器人回包略多于5s企业微信会无回包

+ 配置参数模板语法见: [handlebars][1]
+ 正则表示语法见: [regex][2]

## Developer

1. 下载rust编译环境( https://www.rust-lang.org )
    > 在一些发行版或者软件仓库中也可以通过 pacman/apt/yum/choco 等安装 rust 目标
2. 升级rust工具链 ```rustup self update && rustup update```
3. 安装一个编译目标（默认也会安装一个的） ```rustup target install <目标架构>```
    > 可以通过 ```rustup target list``` 来查看支持的架构
4. 克隆仓库并进入主目录
5. 运行编译命令: ```cargo build```

更多详情见： https://rustup.rs/ 

## LICENSE

[MIT](LICENSE-MIT) or [Apache License - 2.0](LICENSE)

[1]: https://crates.io/crates/handlebars
[2]: https://docs.rs/regex/


CHANGELOG.md
CHANGELOG
============

v0.10.1
----------

1. 修复回包的espace问题
2. 增加 `static_root` 选项，以便用于设置静态文件

v0.10.0
----------

1. 升级 clap 到 ^4.2.0
2. 升级 awc 到 ^3.1.0
3. 升级 rustls 到 >0.21.0
4. 升级 quick-xml 到 >=0.28.0
5. 升级 block-modes 到 >=0.9.0
6. 升级 aes 到 >=0.8.2
7. 升级 cipher 到 >=0.4.0

v0.9.1
----------

1. 升级 actix-web 到 4.0.0-beta.15
2. 升级 tokio 到 1.15.0
3. 使用 awc 发起HTTP请求

v0.9.0
----------

1. 响应环境变量增加 `WXWORK_ROBOT_POST_ID`
2. 命令行发送工具增加发送原始json和模板类消息(灰度功能)的支持

v0.8.0
----------

1. 使用新的Async/Await机制重构代码
2. 移除openssl依赖，现在使用纯rust实现
3. 增加发布包和docker里的script脚本
4. 支持图文混排消息
5. 重新整理输出架构，移除一些不必要的目标，如果同一Target有musl的目标就移除gnu的目标。

v0.7.0
----------

1. 增加对企业微信的新接口的支持（event和attachment）
2. 增加事件类型的回调入口
3. 对配置的 ```type``` 忽略大小写
4. 增加 ```ignore``` 类型事件，可以配置成直接忽略请求
5. 修复默认消息被匹配成空消息的问题

v0.6.2
----------

1. 增加空消息模板配置，用于处理空消息回调
2. 默认情况下空消息不返回错误
3. 修复第一次匹配未跳过隐藏命令的问题
4. 增加 ```order``` 选项，用于控制命令的匹配顺序，默认值为: 0 。越小匹配优先级越高

v0.6.1
----------

1. 更新依赖库
2. 修复有未知字段时解析提前中断的问题
3. 增加对消息的 **ChatType** 参数解析和 **WXWORK_ROBOT_CHAT_TYPE**　环境变量

v0.6.0
----------

1. 更新 [actix-web][1] 更新到新的大版本(^1.0.3)
2. 增加超时时间控制
3. 增加可自定义的消息Body大小配置
4. 增加可自定义连接数的配置
5. 移除对原来 [base64](https://crates.io/crates/base64) 模块的依赖（有BUG，这么高下载量的库实现都有BUG，rust生态真的不太行），改为自己实现的base64算法
6. 增加访问HTTPS的支持

v0.5.1
----------

1. 修订企业微信某些节点不支持"Expect": "100-continue"的问题
2. 调整 UserAgent

v0.5.0
----------

1. 增加一些通用的匹配选项
2. 默认改为大小写不敏感匹配

[1]: https://actix.rs/

src/app.rs
use std::fs::{create_dir_all, OpenOptions};
use std::io::Write;
use std::net::ToSocketAddrs;
use std::path::{Path, PathBuf};
use std::process;
use std::rc::Rc;
use std::str::FromStr;
use std::sync::Arc;

use crate::clap::{Arg, ArgAction, ArgMatches};

use super::wxwork_robot::command::{WxWorkCommandList, WxWorkCommandMatch, WxWorkCommandPtr};
use super::wxwork_robot::project::WxWorkProject;
use super::wxwork_robot::{build_project_set_shared, WxWorkProjectSet, WxWorkProjectSetShared};

#[derive(Debug, Clone)]
pub struct AppConfigure {
    pub task_timeout: u64,
    pub hosts: Option<Vec<String>>,
    pub workers: usize,
    pub backlog: u32,
    pub keep_alive: u64,
    pub client_timeout: u64,
    pub client_shutdown: u64,
    pub max_connection_per_worker: usize,
    pub max_concurrent_rate_per_worker: usize,
    pub payload_size_limit: usize,
    pub static_root: Option<PathBuf>,
}

#[derive(Debug, Clone, Copy)]
pub struct AppEnvironment {
    pub appname: &'static str,
    pub configure: &'static str,
    pub version: &'static str,
    pub prefix: &'static str,
    pub debug: bool,
    pub log: &'static str,
    pub log_rotate: i32,
    pub log_rotate_size: usize,
    pub pid_file: &'static str,
    pub conf: &'static AppConfigure,
}

struct AppEnvironmentInfo {
    init: bool,
    pub debug: bool,
    pub configure: Option<String>,
    pub prefix: Option<String>,
    pub log: Option<String>,
    pub log_rotate: i32,
    pub log_rotate_size: usize,
    pub pid_file: Option<String>,
    pub projects: Option<WxWorkProjectSetShared>,
    pub conf: AppConfigure,
}

static mut APP_ENV_INFO_STORE: AppEnvironmentInfo = AppEnvironmentInfo {
    init: false,
    debug: false,
    configure: None,
    prefix: None,
    log: None,
    log_rotate: 8,
    log_rotate_size: 2097152,
    pid_file: None,
    projects: None,
    conf: AppConfigure {
        task_timeout: 5000,
        hosts: None,
        workers: 8,
        backlog: 256,
        keep_alive: 5,
        client_timeout: 5000,
        client_shutdown: 5000,
        max_connection_per_worker: 20480,
        max_concurrent_rate_per_worker: 256,
        payload_size_limit: 262144, // 256KB
        static_root: None,
    },
};

fn unwraper_flag<S>(matches: &ArgMatches, name: S) -> bool
where
    S: AsRef<str>,
{
    if let Ok(Some(x)) = matches.try_get_one::<bool>(name.as_ref()) {
        return *x;
    }

    false
}

pub trait OptionValueWrapper<T> {
    fn pick(input: &str) -> Option<T>;
}

impl<T> OptionValueWrapper<T> for T
where
    T: FromStr,
{
    fn pick(input: &str) -> Option<Self> {
        if let Ok(v) = input.parse::<T>() {
            Some(v)
        } else {
            None
        }
    }
}

fn unwraper_option<T, S>(matches: &ArgMatches, name: S) -> Option<T>
where
    T: OptionValueWrapper<T>,
    S: AsRef<str>,
{
    if let Ok(Some(x)) = matches.try_get_raw(name.as_ref()) {
        for val in x {
            if let Some(str_val) = val.to_str() {
                return T::pick(str_val);
            }
        }
    }

    None
}

/// Build a clap application parameterized by usage strings.
pub fn app() -> AppEnvironment {
    unsafe {
        if APP_ENV_INFO_STORE.init {
            return generate_app_env();
        }
    }

    let matches = command!();

    let app = matches
        .author(crate_authors!())
        .version(crate_version!())
        .about(crate_description!())
        .max_term_width(120)
        .arg(
            Arg::new("debug")
                .short('d')
                .long("debug")
                .action(ArgAction::SetTrue)
                .help("Show debug log"),
        )
        .arg(
            Arg::new("prefix")
                .short('P')
                .long("prefix")
                .value_name("PREFIX")
                .help("Set a url prefix for current service")
                .default_value("/"),
        )
        .arg(
            Arg::new("configure")
                .short('c')
                .long("conf")
                .value_name("CONFIGURE")
                .help("Set configure file")
                .required(true),
        )
        .arg(
            Arg::new("log")
                .short('l')
                .long("log")
                .value_name("LOG PATH")
                .help("Set log path"),
        )
        .arg(
            Arg::new("log-rotate")
                .long("log-rotate")
                .value_name("LOG ROTATE")
                .help("Set log rotate")
                .default_value("8"),
        )
        .arg(
            Arg::new("log-rotate-size")
                .long("log-rotate-size")
                .value_name("LOG ROTATE SIZE")
                .help("Set log rotate size in bytes"),
        )
        .arg(
            Arg::new("pid-file")
                .long("pid-file")
                .value_name("PID FILE")
                .help("Set path of pid file"),
        );

    let matches: ArgMatches = app.get_matches();

    unsafe {
        if unwraper_flag(&matches, "debug") {
            APP_ENV_INFO_STORE.debug = true;
        }

        if let Some(val) = unwraper_option(&matches, "configure") {
            APP_ENV_INFO_STORE.configure = Some(val);
        }

        if let Some(mut val_str) = unwraper_option::<String, _>(&matches, "prefix") {
            if !val_str.starts_with('/') {
                val_str.insert(0, '/');
            }
            if !val_str.ends_with('/') {
                val_str.push('/');
            }
            APP_ENV_INFO_STORE.prefix = Some(val_str);
        }

        if let Some(val) = unwraper_option(&matches, "log") {
            APP_ENV_INFO_STORE.log = Some(val);
        } else {
            APP_ENV_INFO_STORE.log = Some(format!("{}.log", crate_name!()));
        }

        if let Some(rotate) = unwraper_option(&matches, "log-rotate") {
            APP_ENV_INFO_STORE.log_rotate = rotate;
        }

        if let Some(rotate_size) = unwraper_option(&matches, "log-rotate-size") {
            APP_ENV_INFO_STORE.log_rotate_size = rotate_size;
        }

        if let Some(val) = unwraper_option(&matches, "pid-file") {
            APP_ENV_INFO_STORE.pid_file = Some(val);
        } else {
            APP_ENV_INFO_STORE.pid_file = Some(format!("{}.pid", crate_name!()));
        }

        APP_ENV_INFO_STORE.init = true;
    }

    let app = generate_app_env();

    // write pid
    write_pid_file(app.pid_file);

    app
}

fn write_pid_file(pid_file: &str) {
    // get & create base dir
    let file_path = Path::new(pid_file);
    if let Some(dir_path) = file_path.parent() {
        if !dir_path.as_os_str().is_empty() && (!dir_path.exists() || !dir_path.is_dir()) {
            match create_dir_all(dir_path) {
                Ok(_) => {}
                Err(e) => {
                    eprintln!(
                        "Try to create pid file directory {:?} failed, {}",
                        dir_path, e
                    );
                    return;
                }
            }
        }
    }

    let mut options = OpenOptions::new();
    options.create(true).write(true).truncate(true);
    match options.open(pid_file) {
        Ok(mut file) => match file.write(format!("{}", process::id()).as_bytes()) {
            Ok(_) => {}
            Err(e) => {
                eprintln!(
                    "Try to write {} to pid file {} failed, {}",
                    process::id(),
                    pid_file,
                    e
                );
            }
        },
        Err(e) => {
            eprintln!("Try to open pid file {} failed, {}", pid_file, e);
        }
    }
}

pub fn app_conf() -> &'static AppConfigure {
    let ret;
    unsafe {
        ret = &APP_ENV_INFO_STORE.conf;
    }

    ret
}

fn generate_app_env() -> AppEnvironment {
    unsafe {
        AppEnvironment {
            appname: crate_name!(),
            configure: if let Some(ref x) = APP_ENV_INFO_STORE.configure {
                x.as_str()
            } else {
                "conf.json"
            },
            version: crate_version!(),
            prefix: if let Some(ref x) = APP_ENV_INFO_STORE.prefix {
                x.as_str()
            } else {
                "/"
            },
            debug: APP_ENV_INFO_STORE.debug,
            log: if let Some(ref x) = APP_ENV_INFO_STORE.log {
                x.as_str()
            } else {
                "server.log"
            },
            log_rotate: APP_ENV_INFO_STORE.log_rotate,
            log_rotate_size: APP_ENV_INFO_STORE.log_rotate_size,
            pid_file: if let Some(ref x) = APP_ENV_INFO_STORE.pid_file {
                x.as_str()
            } else {
                "server.pid"
            },
            conf: &APP_ENV_INFO_STORE.conf,
        }
    }
}

impl AppEnvironment {
    fn get_info(&self, is_html: bool) -> String {
        let mut title = format!("{0} {1} Listen on", self.appname, self.version);
        for v in self.get_hosts() {
            title += format!(" \"{0}\"", v).as_str();
        }
        title += format!(" with prefix {0}", self.prefix).as_str();

        let header = if is_html {
            format!(
                "<h3>{}</h3><table><tr><th>Option</th><th>Value</th></tr>",
                title
            )
        } else {
            format!("# {}", title)
        };

        let mut prefix_str = String::from(self.prefix);
        prefix_str.pop();
        let tail = if is_html { "</table>" } else { "" };
        let row_begin = if is_html { "<tr><td>" } else { "" };
        let row_split = if is_html { "</td><td>" } else { ": " };
        let row_end = if is_html { "</td></tr>" } else { "\n" };
        let mut row_host = String::from("");
        for v in self.get_hosts() {
            if let Ok(saddr_iter) = v.to_socket_addrs() {
                for saddr in saddr_iter {
                    if saddr.is_ipv4() {
                        row_host +=
                            format!(
                            "{3}Wechat robot callback URL: {4}http://{0}:{1}{2}/<project name>{5}",
                            saddr.ip(), saddr.port(), prefix_str, row_begin, row_split, row_end
                        )
                            .as_str();
                        if saddr.ip().is_unspecified() {
                            row_host +=
                                format!(
                                "{3}Wechat robot callback URL: {4}http://{0}:{1}{2}/<project name>{5}",
                                "127.0.0.1", saddr.port(), prefix_str, row_begin, row_split, row_end
                            ).as_str();
                        }
                    } else if saddr.is_ipv6() {
                        row_host += format!(
                            "{3}Wechat robot callback URL: {4}http://[{0}]:{1}{2}/<project name>{5}",
                            saddr.ip(), saddr.port(), prefix_str, row_begin, row_split, row_end
                        ).as_str();
                        if saddr.ip().is_unspecified() {
                            row_host +=
                                format!(
                                "{3}Wechat robot callback URL: {4}http://{0}:{1}{2}/<project name>{5}",
                                "::", saddr.port(), prefix_str, row_begin, row_split, row_end
                            ).as_str();
                        }
                    }
                }
            }
        }
        format!("{0}\r\n{1}\r\n{2}", header, row_host, tail)
    }

    pub fn text_info(&self) -> String {
        self.get_info(false)
    }

    pub fn html_info(&self) -> String {
        self.get_info(true)
    }

    pub fn get_hosts(&self) -> Vec<String> {
        let ret: Vec<String>;
        unsafe {
            ret = if let Some(ref x) = APP_ENV_INFO_STORE.conf.hosts {
                x.clone()
            } else {
                Vec::new()
            };
        }

        ret
    }

    pub fn get_projects(&self) -> Option<WxWorkProjectSetShared> {
        let ret: Option<WxWorkProjectSetShared>;
        unsafe {
            ret = APP_ENV_INFO_STORE.projects.as_ref().cloned();
        }

        ret
    }

    pub fn set_projects(&self, val: WxWorkProjectSetShared) {
        {
            if let Ok(x) = val.lock() {
                let ref_x: &WxWorkProjectSet = &x;
                for (k, _) in ref_x.projs.iter() {
                    info!("load project \"{}\" success", k);
                }

                for cmd in ref_x.cmds.iter() {
                    info!("load global command \"{}\" success", cmd.name());
                }

                for cmd in ref_x.events.iter() {
                    info!("load global event \"{}\" success", cmd.name());
                }
            }
        }

        unsafe {
            APP_ENV_INFO_STORE.projects = Some(val);
        }
    }

    pub fn get_project(&self, name: &str) -> Option<Arc<WxWorkProject>> {
        if let Some(projs) = self.get_projects() {
            if let Ok(x) = projs.lock() {
                if let Some(found_proj) = x.projs.get(name) {
                    return Some(found_proj.clone());
                }
            }
        }

        None
    }

    pub fn get_global_command(
        &self,
        message: &str,
        allow_hidden: bool,
    ) -> Option<(WxWorkCommandPtr, WxWorkCommandMatch)> {
        if let Some(projs) = self.get_projects() {
            if let Ok(x) = projs.lock() {
                return WxWorkProject::try_capture_commands(&x.cmds, message, allow_hidden);
            }
        }

        None
    }

    pub fn get_global_event(
        &self,
        message: &str,
        allow_hidden: bool,
    ) -> Option<(WxWorkCommandPtr, WxWorkCommandMatch)> {
        if let Some(projs) = self.get_projects() {
            if let Ok(x) = projs.lock() {
                return WxWorkProject::try_capture_commands(&x.events, message, allow_hidden);
            }
        }

        None
    }

    /// Get global command list.
    ///
    /// **This is a high cost API**
    pub fn get_global_command_list(&self) -> Rc<WxWorkCommandList> {
        if let Some(projs) = self.get_projects() {
            if let Ok(x) = projs.lock() {
                return x.cmds.clone();
            }
        }

        Rc::new(Vec::new())
    }

    pub fn reload(&mut self) -> bool {
        let mut options = OpenOptions::new();
        options.read(true).write(false);
        match options.open(self.configure) {
            Ok(f) => {
                if let Ok(conf) = serde_json::from_reader(f) {
                    let conf_json: serde_json::Value = conf;
                    self.reload_app_conf(&conf_json);
                    if let Some(x) = build_project_set_shared(&conf_json) {
                        self.set_projects(x);
                    } else {
                        error!(
                            "Build project set from configure file {} failed",
                            self.configure
                        );
                        return false;
                    }
                } else {
                    error!("Parse configure file {} as json failed", self.configure);
                    return false;
                }
            }
            Err(e) => {
                error!("Open configure file {} failed, {:?}", self.configure, e);
                return false;
            }
        }

        true
    }

    fn reload_app_conf(&mut self, conf_json: &serde_json::Value) {
        let kvs = if let Some(x) = conf_json.as_object() {
            x
        } else {
            return;
        };

        if let Some(x) = kvs.get("task_timeout") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.task_timeout = v;
                    }
                }
            }
        } else if let Some(x) = kvs.get("taskTimeout") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.task_timeout = v;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("workers") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.workers = v as usize;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("backlog") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.backlog = v as u32;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("keep_alive") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.keep_alive = v;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("client_timeout") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.client_timeout = v;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("client_shutdown") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.client_shutdown = v;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("max_connection_per_worker") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.max_connection_per_worker = v as usize;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("max_concurrent_rate_per_worker") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.max_concurrent_rate_per_worker = v as usize;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("payload_size_limit") {
            if let Some(v) = x.as_u64() {
                if v > 0 {
                    unsafe {
                        APP_ENV_INFO_STORE.conf.payload_size_limit = v as usize;
                    }
                }
            }
        }

        if let Some(x) = kvs.get("static_root") {
            if let Some(v) = x.as_str() {
                if !v.is_empty() {
                    if let Ok(root_path) = PathBuf::from_str(v) {
                        unsafe {
                            APP_ENV_INFO_STORE.conf.static_root = Some(root_path);
                        }
                    }
                }
            }
        }

        {
            let mut hosts = Vec::new();
            if let Some(x) = kvs.get("listen") {
                if let Some(arr) = x.as_array() {
                    for v in arr {
                        if let Some(vh) = v.as_str() {
                            hosts.push(String::from(vh));
                        }
                    }
                } else if let Some(v) = x.as_str() {
                    hosts.push(String::from(v));
                }
            }

            if hosts.is_empty() {
                hosts.push(String::from("0.0.0.0:12019"));
                hosts.push(String::from(":::12019"));
            }

            unsafe {
                APP_ENV_INFO_STORE.conf.hosts = Some(hosts);
            }
        }
    }
}
